File: clients/brpc-client-python/brpc/client/brpc_client.py
'''
Created on 2015-6-7

@author: Administrator
'''

import json
import urllib2

from brpc.bluemei.utils import retry


STATUS_REQUEST = 10
STATUS_REQUEST_LOGIN = 11
STATUS_REQUEST_LOGOUT = 12

STATUS_RESPONSE_OK = 20
STATUS_RESPONSE_EVENT = 26
STATUS_RESPONSE_PENDING = 27
STATUS_RESPONSE_EXCEPTION = 28

STATUS_RESPONSE_ERROR = 30
STATUS_RESPONSE_INVALID_METHOD = 31
STATUS_RESPONSE_INVALID_ARGS = 32
STATUS_RESPONSE_DENIED = 33


class BRpcException(Exception):
    def __init__(self, msg):
        self.message = msg

    def __str__(self):
        return str(self.message)

class BadResponseException(BRpcException):
    def __init__(self, response):
        self.message = str(response)

class BRpcClient(object):
    '''brpc client for python'''

    def __init__(self, url, username, passwd, option=None):
        self.url = url
        self.username = username
        self.passwd = passwd
        self.options = option or {}
        self.authToken = None

        self.login(username, passwd)

    @staticmethod
    def _do_http_post(url, data, options):
        json_data = json.dumps(data)
        timeout = options.get('timeout', 3)
        content_type = 'application/json'  # 'text/xml'
        content_encoding = 'utf-8'

        req = urllib2.Request(url)
        req.add_header('Content-type', content_type)
        req.add_header('Content-encoding', content_encoding)
        req.add_data(json_data)

        @retry.retry_count(3, urllib2.socket.timeout)
        def _do_req(req, timeout):
            response = urllib2.urlopen(req, timeout=timeout)
            return response.read()

        # print '>>>>', url
        # print '>>>>', json_data
        response = _do_req(req, timeout)
        # print '<<<<', response
        return json.loads(response)

    def _do_request(self, data, options=None):
        '''send http post'''
        url = self.url
        if options is None:
            options = self.options
        return self._do_http_post(url, data, options)

    def call(self, method, args, cast=False):
        '''invoke method'''
        body = {
            'method': method,
            'args': args,
            'authToken': self.authToken,
            'needReturn': not cast,
            'status': STATUS_REQUEST,
            'version': '1.0',
            'owner': '',
            'returnValue': None
        }
        response = self._do_request(body)
        status = response['status']
        if cast:
            if status != STATUS_RESPONSE_PENDING:
                raise BadResponseException(response)
        else:
            if status != STATUS_RESPONSE_OK:
                raise BadResponseException(response)
        return response['returnValue']


    def login(self, username, passwd):
        '''login'''
        body = {
            'method': 'login',
            'args': [username, passwd],
            'authToken': self.authToken,
            'needReturn': True,
            'status': STATUS_REQUEST_LOGIN,
            'version': '1.0',
            'owner': ''
        }
        response = self._do_request(body)
        if response['status'] != STATUS_RESPONSE_OK:
            raise BadResponseException(response)
        self.authToken = response['authToken']

    def logout(self):
        '''logout'''
        body = {
            'method': 'logout',
            'args': None,
            'authToken': self.authToken,
            'needReturn': True,
            'status': STATUS_REQUEST_LOGOUT,
            'version': '1.0',
            'owner': ''
        }
        response = self._do_request(body)
        if response['status'] != STATUS_RESPONSE_OK:
            raise BadResponseException(response)
        self.authToken = response['authToken']

    def has_login(self):
        '''is login'''
        return self.authToken is not None

File: clients/brpc-client-java/src/brpc/client/BrpcClient.java
/**
 *
 */
package brpc.client;

import java.util.Arrays;
import java.util.List;

import net.sf.json.JSONObject;
import brpc.client.trans.http.HttpRequest;
import brpc.client.trans.http.HttpResponse;
import brpc.common.exceptions.BadResponseException;
import brpc.common.types.BrpcMethod;
import brpc.common.types.ObjectList;
import brpc.common.utils.C;

/**
 * Brpc Client
 * @author Administrator
 * @date 2015-8-31
 */
public class BrpcClient {

	private final String url;
	private String authToken;

	/**
	 * @param url
	 */
	public BrpcClient(String url) {
        this.url = url;
        this.authToken = null;
	}

	/**
	 * @param url
	 * @param username
	 * @param passwd
	 */
	public BrpcClient(String url, String username, String password) {
		this(url);
        this.login(username, password);
	}

	/**
	 * @summary checkResponse
	 * @author Administrator
	 * @date 2015-9-1
	 * @param result void
	 */
	private void checkResponse(BrpcMethod result) {
		List<Integer> status = Arrays.asList(C.BrpcStatus.STATUS_RESPONSE_OK,
				C.BrpcStatus.STATUS_RESPONSE_PENDING);
		if (!status.contains(result.status)) {
			String msg = result.returnValue != null ?
					result.returnValue.toString() :
					"failed to invoke method: " + result.method;
			throw new BadResponseException(result.status, msg);
		}
	}

	/**
	 * @summary doRequest
	 * @author Administrator
	 * @date 2015-9-1
	 * @param result void
	 */
	protected BrpcMethod doRequest(BrpcMethod body) {
		//to json string
		String data = JSONObject.fromObject(body).toString();
		//do request
		HttpRequest req = new HttpRequest(this.url);
		HttpResponse response = req.doPost(data);
		//to json
		JSONObject json = JSONObject.fromObject(response.getContent());
		//json to object
		BrpcMethod result = (BrpcMethod) JSONObject.toBean(json, BrpcMethod.class);
		checkResponse(result);
		return result;
	}

	/**
	 * @summary call
	 * @author Administrator
	 * @date 2015-8-31
	 * @param object
	 * @param method
	 * @param args
	 * @return Object
	 */
	public Object call(String object, String methodName, ObjectList args) {
		BrpcMethod method = new BrpcMethod();
		method.owner = object;
		method.method = methodName;
		method.args = args;
		method.authToken = this.authToken;
		method.needReturn = true;
		method.status = C.BrpcStatus.STATUS_REQUEST;
		method.version = "1.0";

		BrpcMethod result = this.doRequest(method);
		return result.returnValue;
	}

	/**
	 * @summary cast
	 * @author Administrator
	 * @date 2015-8-31
	 * @param object
	 * @param method
	 * @param args
	 * @return Object
	 */
	public Object cast(String object, String methodName, ObjectList args) {
		BrpcMethod method = new BrpcMethod();
		method.owner = object;
		method.method = methodName;
		method.args = args;
		method.authToken = this.authToken;
		method.needReturn = false;
		method.status = C.BrpcStatus.STATUS_REQUEST;
		method.version = "1.0";

		BrpcMethod result = this.doRequest(method);
		return result.returnValue;
	}

	/**
	 * @summary login
	 * @author Administrator
	 * @date 2015-8-31
	 * @param loginArgs void
	 */
	public void login(String username, String password) {
		BrpcMethod method = new BrpcMethod();
		method.method = "login";
		method.args = new ObjectList(username, password);
		method.authToken = "";
		method.needReturn = true;
		method.status = C.BrpcStatus.STATUS_REQUEST_LOGIN;
		method.version = "1.0";
		method.owner = "";

		BrpcMethod result = this.doRequest(method);
		this.authToken = result.authToken;
	}

	/**
	 * @summary logout
	 * @author Administrator
	 * @date 2015-8-31 void
	 */
	public void logout() {
		BrpcMethod method = new BrpcMethod();
		method.method = "logout";
		method.args = null;
		method.authToken = this.authToken;
		method.needReturn = true;
		method.status = C.BrpcStatus.STATUS_REQUEST_LOGOUT;
		method.version = "1.0";
		method.owner = "";

		BrpcMethod result = this.doRequest(method);
		this.authToken = result.authToken;
	}

	/**
	 * @summary hasLogin
	 * @author Administrator
	 * @date 2015-8-31 void
	 */
	public boolean hasLogin() {
        return this.authToken != null;
	}

}
